### PR TITLE
Open v0.3 API + debiasing

### PR DESCRIPTION
**Pull request by splch**
This new version of the API separates job status and metadata from actual results into two different endpoints.

v0.3 also includes support for error_mitigation settings like symmetrization as described in https://arxiv.org/pdf/2301.07233.pdf

This includes:

* add a new error_mitigation parameter on job submission so users can configure it
* add a new aggregation parameter on results that would allow getting results aggregated under the two different methods described in the paper

### CODE REVIEW

File: pennylane_ionq/device.py
Comment (timmysilv) [inline pennylane_ionq/device.py:143]: we tend to use `UserWarning` because `DeprecationWarning` doesn't appear in some environments. If we just left the target as `qpu`, would this swap happen on the IonQ side? Also, where do you intend on the user seeing this coming from/what motivates the `stacklevel=2`?
```suggestion
                UserWarning,
                stacklevel=2,
```

File: pennylane_ionq/device.py
Comment (timmysilv) [inline pennylane_ionq/device.py:158]: might this already be set from device initialization?

File: pennylane_ionq/device.py
Comment (timmysilv) [inline pennylane_ionq/device.py:218]:
```suggestion
        params = {} if self.sharpen is None else {"sharpen": self.sharpen}
```

File: pennylane_ionq/device.py
Comment (timmysilv) [inline pennylane_ionq/device.py:327]:
```suggestion
        api_key=None,
        error_mitigation=None,
        sharpen=None,
```

File: pennylane_ionq/device.py
Comment (timmysilv) [inline pennylane_ionq/device.py:336]:
```suggestion
            api_key=api_key,
            error_mitigation=error_mitigation,
            sharpen=sharpen,
```

**Review by timmysilv — COMMENTED** [review]
hi @splch, thanks for opening this! it all looks great, I've just left some small comments.

File: pennylane_ionq/api_client.py
Comment (splch) [reply pennylane_ionq/api_client.py:124]: yeah, this reordering is intentional - also PENNYLANE_IONQ_API_KEY is added because sometimes we have different clients interact in different ways in tests, so this just specifies the pennylane client

File: pennylane_ionq/device.py
Comment (splch) [reply pennylane_ionq/device.py:158]: yup! thanks for the catch

Comment (timmysilv) [reply pennylane_ionq/device.py:158]: do you also want to add support for it being passed to `apply` through kwargs, or just remove this line?

Comment (splch) [reply pennylane_ionq/device.py]: probably just remove it

**Comment by timmysilv** [comment]
all looks good, but seems like some tests and formatting checks failed (see the CI from my merge commit). once those are fixed, this PR should be good to go!

Also, don't forget to add your name and a quick PR description in the changelog 😄

**Comment by splch** [comment]
> all looks good, but seems like some tests and formatting checks failed (see the CI from my merge commit). once those are fixed, this PR should be good to go!
> 
> Also, don't forget to add your name and a quick PR description in the changelog 😄

hi! the tests should pass now :)

**Comment by timmysilv** [comment]
bumping the changelog reminder, but otherwise lgtm!

**Comment by splch** [comment]
> bumping the changelog reminder, but otherwise lgtm!

just got it! thanks for the help :)

**Comment by timmysilv** [comment]
looks like tests/formatting are still failing ☹️  You can fix formatting by running `black -l 100 pennylane_ionq`, and the tests can be fixed like this:
```diff
diff --git a/tests/test_api_client.py b/tests/test_api_client.py
index e30d8e8..fed274d 100755
--- a/tests/test_api_client.py
+++ b/tests/test_api_client.py
@@ -303,7 +303,7 @@ class TestResourceManager:
         mock_get_response = MockGETResponse(200)
 
         monkeypatch.setattr(
-            requests, "get", lambda url, timeout, headers: mock_get_response
+            requests, "get", lambda url, params, timeout, headers: mock_get_response
         )
         monkeypatch.setattr(
             requests,
diff --git a/tests/test_device.py b/tests/test_device.py
index 328924b..051c038 100755
--- a/tests/test_device.py
+++ b/tests/test_device.py
@@ -121,7 +121,7 @@ class TestDeviceIntegration:
         )
         monkeypatch.setattr(Job, "is_complete", True)
 
-        def fake_response(self, resource_id=None):
+        def fake_response(self, resource_id=None, params=None):
             """Return fake response data"""
             fake_json = {"histogram": {"0": 1}}
             setattr(
```

Also, can you add one line to the `Improvements` section of the changelog just mentioning that you've bumped the IonQ api version being used?

It all looks great, happy to approve once those changes are in 😄

File: CHANGELOG.md
Comment (trbromley) [inline CHANGELOG.md:18]: @splch, it might be nice to highlight the error mitigation support in the changelog

File: CHANGELOG.md
Comment (timmysilv) [inline CHANGELOG.md:11]: seems like your name got removed from the contributor list below, please feel free to add it back. also just updating this to follow our changelog conventions:
```suggestion
* adds a new `error_mitigation` parameter on IonQ job submission so users can
  configure strategies
  [(#75)](https://github.com/PennyLaneAI/PennyLane-IonQ/pull/75)

* adds a new `aggregation` parameter on IonQ results to allow getting results
  aggregated under the two different methods described in
  <https://arxiv.org/pdf/2301.07233.pdf>
  [(#75)](https://github.com/PennyLaneAI/PennyLane-IonQ/pull/75)
```

File: pennylane_ionq/device.py
Comment (timmysilv) [inline pennylane_ionq/device.py:135]: Codecov is upset that this line isn't being tested. updating a test to pass a value to this kwarg should be enough.

File: pennylane_ionq/api_client.py
Comment (timmysilv) [inline pennylane_ionq/api_client.py:272]: codecov is also complaining about this line, but I'm not entirely sure why - it was already here before. if you can test it, that would be sweet. otherwise I'm sure we can figure out a way around it for this PR

**Review by timmysilv — COMMENTED** [review]
codecov seems upset with this change, so CI is still failing. Just adding a bit more test coverage for the new lines should do it, but I don't want to hold this PR up any longer. lmk if there's any trouble adding that in, and we can otherwise get this merged!

**Comment by lillian542** [comment]
Hi @splch! We’ll be releasing an updated version of this plugin as part of the PennyLane release at the beginning of January, so I’m going to go ahead and wrap up the remaining testing and docstrings things and get this PR merged. 

I'll get the changes in this afternoon, and then if you have any edits you would like to make, please add them by Friday. Thanks so much for your work on this!

File: pennylane_ionq/api_client.py
Comment (lillian542) [reply pennylane_ionq/api_client.py:272]: coverage added

File: pennylane_ionq/device.py
Comment (lillian542) [reply pennylane_ionq/device.py:144]: coverage added

File: pennylane_ionq/device.py
Comment (lillian542) [reply pennylane_ionq/device.py]: updated

File: CHANGELOG.md
Comment (lillian542) [reply CHANGELOG.md]: @trbromley, combined the two kwargs into a little blurb and moved to "new features", added a link to an IonQ guide for more info

File: CHANGELOG.md
Comment (lillian542) [inline CHANGELOG.md:6]:
```suggestion
  need to be set at runtime are defined in the `error_mitigation` dictionary (currently a single strategy, `debias`, is available). Whether or not to
```

**Review by mudit2812 — APPROVED** [review]

File: pennylane_ionq/device.py
Comment (timmysilv) [inline pennylane_ionq/device.py:341]: changed the default `backend` to "harmony" because every job submission was raising a warning unless the user specifies "harmony" (or some other backend). Since I believe this was effectively the default before, this will keep that behaviour and not give users warnings (unless they specify `None`, in which case the warning is justified)

**Pull request merged**
